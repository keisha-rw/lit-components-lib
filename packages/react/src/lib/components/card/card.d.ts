import * as React from 'react';
export declare const PdsCard: React.FunctionComponent<Partial<{
    ariaLabel: string;
    target: string | undefined;
    variant: "default" | "bare";
    href: string | undefined;
    centerContentVertically: boolean;
    borderRadiusSize: "default" | "sm";
    direction: "default" | "horizontal";
    removePadding: boolean;
    hasNoSlottedLinkContent: boolean;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{
    onClick: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
