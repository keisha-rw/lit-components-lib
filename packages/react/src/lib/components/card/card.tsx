import * as React from 'react';
import { PdsCard as PdsCardElement } from '@keisha/design-system';
import { EventName } from '@lit-labs/react';
import { PdsCardProps } from '../../built-component-props';
import { createPdsReactComponent } from '../../create-pds-react-component';

export const PdsCard = createPdsReactComponent<PdsCardProps>()({
  tagName: 'pds-card',
  elementClass: PdsCardElement,
  react: React,
  events: {
    onClick: 'pds-card-click' as EventName<CustomEvent>,
  },
});
