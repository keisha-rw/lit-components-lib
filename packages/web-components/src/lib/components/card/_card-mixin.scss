/*------------------------------------*\
 #CARD_MIXIN
\*------------------------------------*/

@mixin pds-c-card {
  border: 1px solid var(--pds-semantic-border-default);
  border-radius: size(1);
  height: 100%;
  overflow: hidden;
}
