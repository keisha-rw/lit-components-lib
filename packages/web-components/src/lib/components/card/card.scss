@import './card-mixin';
@import '../link/link-mixin';

/*------------------------------------*\
 #CARD
\*------------------------------------*/

:host {
  background-color: var(--pds-semantic-background-default);
  border-radius: size(1);
  display: block;
  height: 100%;
}

:host([variant='bare']) {
  background-color: transparent;
}

.pds-c-card {
  @include pds-c-card;

  display: flex;
  flex-direction: column;
}

.pds-c-card--border-radius-sm {
  border-radius: size(0.5);
}

.pds-c-card--bare {
  border: none;
}

.pds-c-card--clickable {
  display: block;
  position: relative;
  text-decoration: none;
  border: none;
  box-shadow: 0 0 size(1.5) var(--pds-semantic-border-default); // TODO: Change to shadow token
  cursor: pointer;
}

.pds-c-card--center-vertically.pds-c-card--clickable {
  article {
    @include pds-c-card;

    display: flex;
    flex-direction: column;
  }
}

.pds-c-card.pds-c-card--horizontal {
  flex-direction: unset;
}

.pds-c-card__header {
  padding: 0;

  ::slotted(div) {
    overflow: hidden;
  }

  .pds-c-card--horizontal & {
    flex: 0;
  }

  .pds-c-card--bare & {
    ::slotted(div) {
      border-radius: size(1);
    }
  }
}

.pds-c-card__body {
  flex: 1; // Stretches the body for vertical cards in grids or content in horizontal cards
  padding: size(3);

  .pds-c-card--remove-padding & {
    padding: 0;
  }

  .pds-c-card--center-vertically & {
    display: grid;
    align-items: center;
    vertical-align: middle;
  }

  .pds-c-card--bare & {
    padding: size(3) 0;
  }

  .pds-c-card--bare.pds-c-card--remove-padding & {
    padding: 0;
  }

  .pds-c-card--bare.pds-c-card--horizontal & {
    padding: 0 size(3);
  }

  .pds-c-card--bare.pds-c-card--horizontal.pds-c-card--remove-padding & {
    padding: 0;
  }
}

.pds-c-card__footer {
  padding: 0 size(3) size(3);

  .pds-c-card--remove-padding & {
    padding: 0;
  }

  .pds-c-card--bare & {
    padding: 0;
  }

  .pds-c-card--horizontal & {
    padding: size(3);
    margin-left: auto;
  }

  .pds-c-card--horizontal.pds-c-card--remove-padding & {
    padding: 0;
  }

  .pds-c-card--bare.pds-c-card--horizontal & {
    padding: size(3) 0;
  }

  .pds-c-card--bare.pds-c-card--horizontal.pds-c-card--remove-padding & {
    padding: 0;
  }
}

.pds-c-card--horizontal {
  display: flex;
}
