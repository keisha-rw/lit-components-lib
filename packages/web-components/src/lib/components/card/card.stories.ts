import { Meta, StoryObj } from '@storybook/web-components';
import { html, nothing } from 'lit';
import './card';
import '../../../../.storybook/components/placeholder-element/placeholder-element';
import '../link/link';

export default {
  title: 'Components/Card',
  component: 'pds-card',
  tags: ['autodocs'],
  argTypes: {
    variant: {
      control: 'select',
      options: ['default', 'bare'],
    },
    borderRadiusSize: {
      control: 'select',
      options: ['default', 'sm'],
    },
    direction: {
      control: 'select',
      options: ['default', 'horizontal'],
    },
    href: {
      control: 'text',
    },
    target: {
      control: 'text',
    },
  },
  render: (args) =>
    html`<pds-card
      variant="${args.variant || nothing}"
      href="${args.href || nothing}"
      centerContentVertically="${args.centerContentVertically || nothing}"
      target="${args.target || nothing}"
      removePadding="${args.removePadding || nothing}"
      ariaLabel="${args.ariaLabel || nothing}"
      direction="${args.direction || nothing}"
      borderRadiusSize="${args.borderRadiusSize || nothing}"
    >
      <div slot="header">
        <placeholder-element style="--placeholder-element-margin-bottom: 0;"
          >Card header slot</placeholder-element
        >
      </div>
      <placeholder-element style="--placeholder-element-margin-bottom: 0;"
        >Card body slot</placeholder-element
      >
      <div slot="footer">
        <placeholder-element style="--placeholder-element-margin-bottom: 0;"
          >Card footer slot</placeholder-element
        >
      </div>
    </pds-card>`,
  parameters: {
    actions: {
      handles: ['pds-card-click', 'pds-link-click'],
    },
  },
} as Meta;

export const Default: StoryObj = {
  name: 'Default',
  args: {},
};

export const Bare: StoryObj = {
  name: 'Bare',
  args: {
    variant: 'bare',
  },
};

export const Clickable: StoryObj = {
  name: 'Clickable',
  args: {
    href: 'https://www.google.com',
    target: '_blank',
    ariaLabel: 'This is an aria label',
  },
  render: (args) =>
    html`<pds-card
      href="${args.href}"
      target="${args.target || nothing}"
      ariaLabel="${args.ariaLabel || nothing}"
      centerContentVertically="${args.centerContentVertically || nothing}"
    >
      <div slot="header">
        <placeholder-element style="--placeholder-element-margin-bottom: 0;"
          >Card header slot</placeholder-element
        >
      </div>
      <div>
        <pds-link href="https://www.google.com" target="_blank"
          >Example link</pds-link
        >
      </div>
      <div slot="footer">
        <placeholder-element style="--placeholder-element-margin-bottom: 0;"
          >Card footer slot</placeholder-element
        >
      </div>
    </pds-card>`,
};

export const ClickableNoInternalLink: StoryObj = {
  name: 'Clickable no internal link',
  args: {
    target: '_blank',
    href: `https://www.google.com/individuals/invest-retire/individual-retirement-account?alert('xss')`,
    ariaLabel: 'This is an aria label',
  },
};

export const Horizontal: StoryObj = {
  name: 'Horizontal',
  args: { direction: 'horizontal' },
};

export const SmallBorderRadius: StoryObj = {
  name: 'Small border radius',
  args: { borderRadiusSize: 'sm' },
};

export const RemovePadding: StoryObj = {
  name: 'Remove default padding',
  args: { removePadding: true },
};

export const HorizontalWithoutPadding: StoryObj = {
  name: 'Horizontal without padding',
  args: { removePadding: true, direction: 'horizontal' },
  render: (args) =>
    html`<pds-card
      variant="${args.variant || nothing}"
      href="${args.href || nothing}"
      centerContentVertically="${args.centerContentVertically || nothing}"
      target="${args.target || nothing}"
      removePadding="${args.removePadding || nothing}"
      ariaLabel="${args.ariaLabel || nothing}"
      direction="${args.direction || nothing}"
      borderRadiusSize="${args.borderRadiusSize || nothing}"
    >
      <div slot="header">
        <placeholder-element style="--placeholder-element-margin-bottom: 0;"
          >Card header slot</placeholder-element
        >
      </div>
      <placeholder-element
        inlineStyle="height: 118px; display: flex; align-items: center; justify-content: center;"
        ><p>Card body slot</p></placeholder-element
      >
      <div slot="footer">
        <placeholder-element
          inlineStyle="height: 118px; display: flex; align-items: center; justify-content: center;"
          >Card footer slot</placeholder-element
        >
      </div>
    </pds-card>`,
};

export const HorizontalWithUtilClasses: StoryObj = {
  name: 'Horizontal with util class',
  render: () =>
    html`<pds-card direction="horizontal" class="pds-u-background-subtle">
      <div slot="header">
        <placeholder-element style="--placeholder-element-margin-bottom: 0;"
          >Card header slot</placeholder-element
        >
      </div>
      <placeholder-element
        inlineStyle="height: 118px; display: flex; align-items: center; justify-content: center;"
        ><p>Card body slot</p></placeholder-element
      >
      <div slot="footer">
        <placeholder-element
          inlineStyle="height: 118px; display: flex; align-items: center; justify-content: center;"
          >Card footer slot</placeholder-element
        >
      </div>
    </pds-card>`,
};
