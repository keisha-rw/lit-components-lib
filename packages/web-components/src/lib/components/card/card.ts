import { html, nothing } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { sanitizeUrl } from '@braintree/sanitize-url';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsElement } from '../PdsElement';
import styles from './card.scss?inline';

/**
 * @summary This component creates a card container for other elements
 *
 * @slot header This slot holds the header content of the card
 * @slot default This is the body content of the card
 * @slot footer This slot holds the footer content of the card
 *
 * @fires pds-card-click A custom event dispatched on a card click when href is provided
 */
@customElement('pds-card', {
  category: 'component',
  type: 'component',
  state: 'stable',
  styles,
})
export class PdsCard extends PdsElement {
  /**
   * - **default** renders a card with a subtle border and padding
   * - **bare** renders a card without a border or drop shadow
   */
  @property()
  variant: 'default' | 'bare' = 'default';

  /**
   * - **default** renders a card with a border radius of 8px
   * - **sm** renders a card with a border radius of 4px
   */
  @property()
  borderRadiusSize: 'sm' | 'default' = 'default';

  /**
   * Adds an additional class for aligning items to the vertical center of the card.
   */
  @property({ type: Boolean })
  centerContentVertically: boolean = false;

  /**
   * - **default** renders a card in a vertical format
   * - **horizontal** renders a card in a horizontal format
   */
  @property()
  direction: 'horizontal' | 'default' = 'default';

  /**
   * Provides the src value for the href card, renders a clickable card
   * styled with a drop shadow but no border
   */
  @property()
  href?: string;

  /**
   * Used in combination with the href property to determine the target
   * for the clickable card. Not providing a value opens link in the same tab.
   * For a complete list of valid strings, see https://html.spec.whatwg.org/multipage/document-sequences.html#valid-navigable-target-name-or-keyword
   */
  @property()
  target?: string;

  /**
   * ARIA label string that describes a clickable pds-card for screen reader users
   */
  @property()
  ariaLabel: string;

  /**
   * If true, this property will remove the default padding from the card and
   * the content will extend all the way to the edges
   */
  @property({ type: Boolean })
  removePadding: boolean = false;

  /**
   * Boolean to represent that a card has no slotted link content, used
   * to determine logic for a11y features of the clickable card
   *
   * @internal
   */
  @property({ type: Boolean })
  hasNoSlottedLinkContent: boolean = true;

  /**
   * This grabs the pds-links from the header slot
   * @internal
   */
  @queryAssignedElements({ slot: 'header' })
  headerSlotElements: HTMLElement[];

  /**
   * This grabs the pds-links from the default slot
   * @internal
   */
  @queryAssignedElements({ slot: undefined })
  defaultSlotElements: HTMLElement[];

  /**
   * This grabs the pds-links from the footer slot
   * @internal
   */
  @queryAssignedElements({ slot: 'footer' })
  footerSlotElements: HTMLElement[];

  /**
   * Get summary text
   *
   * @returns string summary text
   *
   * @internal
   */
  getSummaryText(textContent: string | null) {
    if (textContent) {
      return textContent.trim();
    }

    return '';
  }

  /**
   * On click, a clickable card will open the src link and fire an event
   *
   * @internal
   */
  handleClick(e: MouseEvent) {
    if (this.href) {
      const target = e.target as HTMLElement;
      // fire event if a clickable card is clicked but the click isn't coming from an enclosed pds-link or anchor element
      if (
        (target.tagName.toLowerCase() !== 'pds-link' &&
          target.tagName.toLowerCase() !== 'a') ||
        target.classList.contains(this.classMod('clickable'))
      ) {
        const customEvent = new CustomEvent('pds-card-click', {
          bubbles: true,
          composed: true,
          detail: {
            summary: this.getSummaryText(this.textContent),
          },
        });
        this.dispatchEvent(customEvent);
        if (window) {
          if (this.target) {
            window.open(this.href, this.target);
          } else {
            window.location.href = sanitizeUrl(
              this.href ?? window.location.href,
            );
          }
        }
      }
    }
  }

  handleKeydown(e: KeyboardEvent) {
    if (this.href) {
      if (e.key === 'Enter') {
        // fire event if a clickable card is clicked with 'Enter' Key
        const customEvent = new CustomEvent('pds-card-click', {
          bubbles: true,
          composed: true,
          detail: {
            summary: this.getSummaryText(this.textContent),
          },
        });
        this.dispatchEvent(customEvent);
        if (window) {
          if (this.target) {
            window.open(this.href, this.target);
          } else {
            window.location.href = sanitizeUrl(
              this.href ?? window.location.href,
            );
          }
        }
      }
    }
  }

  /**
   * On mouseover, a clickable card will add the attribute 'hover' on
   * elements with a class of .pds-c-card__element-with-hover-state
   *
   * @internal
   */
  handleMouseover() {
    if (this.href) {
      const setHover = (element: HTMLElement) => {
        if (
          element.classList.contains(this.classEl('element-with-hover-state'))
        ) {
          element.setAttribute('hover', '');
        } else {
          element
            .querySelectorAll(`.${this.classEl('element-with-hover-state')}`)
            .forEach((hoverableItem) => {
              hoverableItem.setAttribute('hover', '');
            });
        }
      };

      this.headerSlotElements.forEach((element) => setHover(element));
      this.defaultSlotElements.forEach((element) => setHover(element));
      this.footerSlotElements.forEach((element) => setHover(element));
    }
  }

  /**
   * On mouseout, a clickable card will remove the attribute 'hover' on
   * elements with a class of .pds-c-card__element-with-hover-state
   */
  handleMouseout() {
    if (this.href) {
      const removeHover = (element: HTMLElement) => {
        if (
          element.classList.contains(this.classEl('element-with-hover-state'))
        ) {
          element.removeAttribute('hover');
        } else {
          element
            .querySelectorAll(`.${this.classEl('element-with-hover-state')}`)
            .forEach((hoverableItem) => {
              hoverableItem.removeAttribute('hover');
            });
        }
      };

      this.headerSlotElements.forEach((element: HTMLElement) =>
        removeHover(element),
      );
      this.defaultSlotElements.forEach((element: HTMLElement) =>
        removeHover(element),
      );
      this.footerSlotElements.forEach((element: HTMLElement) =>
        removeHover(element),
      );
    }
  }

  /**
   * This function checks for <pds-link> and <a> tags within the card,
   * (first as the only slotted item, then as a child of a slotted item)
   * and then adds the hover class that is targeted to provide the hover
   * attribute when a clickable card is hovered.
   *
   * The <pds-link> tag will handle the hover effect due to the built
   * in support for the 'hover' attribute. An <a> tag will need to have
   * the hover effect coded by the consuming application.
   */
  handleSlotChange(e: { target: { assignedElements: () => any } }) {
    if (this.href) {
      const childElements = e.target.assignedElements();
      childElements.forEach((element: HTMLElement) => {
        if (
          element.tagName.toLowerCase() === 'pds-link' ||
          element.tagName.toLowerCase() === 'a'
        ) {
          this.hasNoSlottedLinkContent = false;
          element.classList.add(this.classEl('element-with-hover-state'));
        } else {
          element.querySelectorAll('pds-link, a').forEach((internalElement) => {
            this.hasNoSlottedLinkContent = false;
            internalElement.classList.add(
              this.classEl('element-with-hover-state'),
            );
          });
        }
      });
    }
  }

  preflight() {
    if (this.target && !this.href) {
      console.warn(
        'A target property without an href property will have no effect. Remove the target property to remove this warning.',
        this,
      );
    }
    if (this.ariaLabel && !this.href) {
      console.warn(
        'An ariaLabel property without an href property will have no effect. Remove the ariaLabel property to remove this warning.',
        this,
      );
    }
    if (this.direction === 'horizontal' && this.centerContentVertically) {
      console.warn(
        'The centerContentVertically property cannot be used when the card has a horizontal direction. Remove the centerContentVertically property to remove this warning.',
        this,
      );
    }
    if (this.variant === 'bare') {
      this.setAttribute('variant', 'bare');
    }
  }

  /**
   * @internal
   */
  get classNames() {
    return {
      [`${this.variant}`]: !!this.variant,
      [`${this.direction}`]: !!this.direction,
      [`border-radius-${this.borderRadiusSize}`]: !!this.borderRadiusSize,
      clickable: !!this.href,
      'center-vertically':
        this.direction !== 'horizontal' &&
        this.centerContentVertically === true,
      'remove-padding': !!this.removePadding,
    };
  }

  willUpdate() {
    if (
      this.innerHTML.indexOf('pds-link') > -1 ||
      this.innerHTML.indexOf('<a') > -1
    ) {
      this.hasNoSlottedLinkContent = false;
    }
  }

  updated(): void {
    if (this.shadowRoot) {
      const clickableAnchor = this.shadowRoot.querySelector(
        `.${this.classMod('clickable')}`,
      );
      // Make entire card link tabbable only if there is no visible internal link
      if (clickableAnchor && this.hasNoSlottedLinkContent) {
        clickableAnchor.setAttribute('tabindex', '0');
      }
    }
  }

  render() {
    this.preflight();
    if (this.href && this.hasNoSlottedLinkContent && !this.ariaLabel) {
      console.error(
        'Cards with a href but no slotted pds-link must include an aria label.',
        this,
      );
      return nothing;
    }
    return html`<article
      class=${this.getClass()}
      @click=${this.handleClick}
      @keydown=${this.handleKeydown}
      @mouseover=${this.handleMouseover}
      @mouseout=${this.handleMouseout}
    >
      ${this.slotNotEmpty('header') &&
      html`
        <header class="${this.classEl('header')}">
          <slot @slotchange=${this.handleSlotChange} name="header"></slot>
        </header>
      `}
      <div class="${this.classEl('body')}">
        <slot @slotchange=${this.handleSlotChange}></slot>
      </div>
      ${this.slotNotEmpty('footer') &&
      html`
        <footer class="${this.classEl('footer')}">
          <slot @slotchange=${this.handleSlotChange} name="footer"></slot>
        </footer>
      `}
    </article>`;
  }
}
