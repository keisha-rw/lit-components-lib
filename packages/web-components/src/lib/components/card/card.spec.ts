import * as path from 'path';
import initStoryshots from '@storybook/addon-storyshots';
import { assert, fixture } from '@open-wc/testing';
import userEvent from '@testing-library/user-event';
import { PdsCard } from './card';
import createSnapshots from '../../../../test/utils/snapshots';

initStoryshots({
  storyKindRegex: /Card/,
  framework: 'web-components',
  configPath: path.join(__dirname, '../../../../.storybook'),
  asyncJest: true,
  test: createSnapshots,
});

describe('Card unit tests', () => {
  let element: Element;

  beforeEach(async () => {
    element = await fixture(
      '<pds-card target="_blank" href="https://www.google.com" ariaLabel="testing"><span>This is a PDS card</span></pds-card>',
    );
  });

  it('is an instance of PdsCard', () => {
    expect(element).toBeInstanceOf(PdsCard);
  });

  it('populates the header slot when provided', async () => {
    element = await fixture(
      '<pds-card><div slot="header">header text and stuff</div></pds-card>',
    );
    expect(element).toMatchSnapshot();
  });

  it('populates the footer slot when provided', async () => {
    element = await fixture(
      '<pds-card><div slot="footer">Footer text and stuff</div></pds-card>',
    );
    expect(element).toMatchSnapshot();
  });

  it('is accessible', async () => {
    await assert.isAccessible(element);
  });

  it('fires the pds-card-click event if a clickable card is clicked', async () => {
    const handlePdsCardClick = jest.fn();

    const configuredElement = (await fixture(
      `<pds-card target="_blank" href="https://www.google.com" ariaLabel="Test label">
      <span id="test">This is a PDS card</span></pds-card>`,
    )) as PdsCard;
    configuredElement.addEventListener('pds-card-click', handlePdsCardClick);

    const userEventWithoutDelay = userEvent.setup({ delay: null });
    // Fire a click event on the card
    await userEventWithoutDelay.click(
      configuredElement.shadowRoot?.querySelector('article')!,
    );

    expect(handlePdsCardClick).toHaveBeenCalledTimes(1);
  });

  it('fires the pds-card-click event if a clickable card is clicked without a target', async () => {
    const handlePdsCardClick = jest.fn();

    const configuredElement = (await fixture(
      `<pds-card href="#">
      <span id="test"><pds-link href="#">This is a PDS card</pds-link></span></pds-card>`,
    )) as PdsCard;
    configuredElement.addEventListener('pds-card-click', handlePdsCardClick);

    const userEventWithoutDelay = userEvent.setup({ delay: null });
    // Fire a click event on the card
    await userEventWithoutDelay.click(
      configuredElement.shadowRoot?.querySelector('article')!,
    );

    expect(handlePdsCardClick).toHaveBeenCalledTimes(1);
  });

  it('fires the pds-card-click event if a clickable card is clicked and there is no target', async () => {
    const handlePdsCardClick = jest.fn();

    const configuredElement = (await fixture(
      `<pds-card href="https://www.google.com" ariaLabel="Test label">
      <span id="test">This is a PDS card</span></pds-card>`,
    )) as PdsCard;
    const articleElement =
      configuredElement.shadowRoot?.querySelector('article')!;
    configuredElement.addEventListener('pds-card-click', handlePdsCardClick);

    const userEventWithoutDelay = userEvent.setup({ delay: null });

    // Fire a click event on the card
    await userEventWithoutDelay.click(articleElement);

    expect(handlePdsCardClick).toHaveBeenCalledTimes(1);
  });

  it('fires the pds-card-click event if a clickable card is clicked with Enter Key', async () => {
    const pdsCardClickOnKeydownHandler = jest.fn();

    const testFixture = (await fixture(
      `<pds-card target="_blank" href="https://www.google.com" ariaLabel="Test label">
      <span id="test">This is a PDS card</span></pds-card>`,
    )) as PdsCard;
    const articleElement = testFixture.shadowRoot?.querySelector('article')!;
    testFixture.addEventListener(
      'pds-card-click',
      pdsCardClickOnKeydownHandler,
    );

    const userEventWithoutDelay = userEvent.setup({ delay: null });
    // Focus on the card
    articleElement.focus();
    // Hit enter while focused on the card
    await userEventWithoutDelay.keyboard('{enter}');

    expect(pdsCardClickOnKeydownHandler).toHaveBeenCalledTimes(1);
  });

  it('fires the pds-card-click event and sets window location if a clickable card is clicked with Enter Key without a target', async () => {
    const pdsCardClickOnKeydownHandler = jest.fn();

    const testFixture = (await fixture(
      `<pds-card href="www.google.com" ariaLabel="test">
      <span href="#">This is a PDS card</span></pds-card>`,
    )) as PdsCard;
    const articleElement = testFixture.shadowRoot?.querySelector('article')!;
    testFixture.addEventListener(
      'pds-card-click',
      pdsCardClickOnKeydownHandler,
    );

    const userEventWithoutDelay = userEvent.setup({ delay: null });
    // Focus on the card
    articleElement.focus();
    // Hit enter while focused on the card
    await userEventWithoutDelay.keyboard('{enter}');

    expect(pdsCardClickOnKeydownHandler).toHaveBeenCalledTimes(1);
  });

  it('does not fire the pds-card-click event if a clickable card is clicked with space Key', async () => {
    const pdsCardClickOnKeydownHandler = jest.fn();

    const testFixture = (await fixture(
      `<pds-card target="_blank" href="https://www.google.com" ariaLabel="Test label">
      <span id="test">This is a PDS card</span></pds-card>`,
    )) as PdsCard;
    const articleElement = testFixture.shadowRoot?.querySelector('article')!;
    testFixture.addEventListener(
      'pds-card-click',
      pdsCardClickOnKeydownHandler,
    );

    const userEventWithoutDelay = userEvent.setup({ delay: null });
    // Focus on the card
    articleElement.focus();
    // Hit enter while focused on the card
    await userEventWithoutDelay.keyboard(' ');

    expect(pdsCardClickOnKeydownHandler).not.toHaveBeenCalled();
  });

  it('fires the pds-card-click event if a clickable card is clicked with Enter Key and there is no target', async () => {
    const handlePdsCardKeydown = jest.fn();

    const testFixture = (await fixture(
      `<pds-card href="https://www.google.com" ariaLabel="Test label">
      <span id="test">This is a PDS card</span></pds-card>`,
    )) as PdsCard;
    const articleElement = testFixture.shadowRoot?.querySelector('article')!;
    testFixture.addEventListener('pds-card-click', handlePdsCardKeydown);

    const userEventWithoutDelay = userEvent.setup({ delay: null });
    // Focus on the card
    articleElement.focus();
    // Hit enter while focused on the card
    await userEventWithoutDelay.keyboard('{Enter}');

    expect(handlePdsCardKeydown).toHaveBeenCalledTimes(1);
  });

  it('sets hover attribute if a mouseover event occurs on clickable card', async () => {
    const configuredElement = (await fixture(
      `<pds-card href="#">
      <div slot="header"><pds-link href="#" id="testThree" class="pds-c-card__element-with-hover-state">
      This is a link</pds-link></div>
      <span id="test" class="pds-c-card__element-with-hover-state">This is a PDS card</span>
      <div slot="footer"><span id="testTwo" class="pds-c-card__element-with-hover-state">
      This is a PDS card</span></div>
      </pds-card>`,
    )) as PdsCard;

    configuredElement.handleMouseover();

    expect(
      configuredElement.querySelector('#test')?.getAttribute('hover'),
    ).toEqual('');

    expect(
      configuredElement.querySelector('#testTwo')?.getAttribute('hover'),
    ).toEqual('');

    expect(
      configuredElement.querySelector('#testThree')?.getAttribute('hover'),
    ).toEqual('');
  });

  it('does not set the hover attribute if a mouseover event occurs on a card without the href prop', async () => {
    const configuredElement = (await fixture(
      `<pds-card>
      <div slot="header"><pds-link id="testThree" class="pds-c-card__element-with-hover-state">
      This is a link</pds-link></div>
      <span id="test" class="pds-c-card__element-with-hover-state">This is a PDS card</span>
      <div slot="footer"><span id="testTwo" class="pds-c-card__element-with-hover-state">
      This is a PDS card</span></div>
      </pds-card>`,
    )) as PdsCard;

    configuredElement.handleMouseover();

    expect(
      configuredElement.querySelector('#test')?.getAttribute('hover'),
    ).toBeFalsy();
    expect(
      configuredElement.querySelector('#testTwo')?.getAttribute('hover'),
    ).toBeFalsy();
    expect(
      configuredElement.querySelector('#testThree')?.getAttribute('hover'),
    ).toBeFalsy();
  });

  it('removes hover attribute if a mouseout event occurs on clickable card', async () => {
    const configuredElement = (await fixture(
      '<pds-card href="www.google.com" ariaLabel="Test label"><div slot="header"><span hover="" id="testTwo" class="pds-c-card__element-with-hover-state">This is a PDS card</span></div><span hover="" id="test" class="pds-c-card__element-with-hover-state">This is a PDS card</span><div slot="footer"><a href="#" id="testThree" hover>This is an anchor tag link</a></div></pds-card>',
    )) as PdsCard;

    configuredElement.handleMouseout();

    expect(
      configuredElement.querySelector('#test')?.getAttribute('hover'),
    ).toBeFalsy();
    expect(
      configuredElement.querySelector('#testTwo')?.getAttribute('hover'),
    ).toBeFalsy();
    expect(
      configuredElement.querySelector('#testThree')?.getAttribute('hover'),
    ).toBeFalsy();
  });

  it('slot change updates hasNoSlottedLinkContent and adds proper class for clickable card when pds-link is included in a slot', async () => {
    const configuredElement = (await fixture(
      `<pds-card href="#" ariaLabel="Test label"><div slot="header">
      <pds-link href="#" id="test" class="pds-c-card__element-with-hover-state">This is a PDS link</pds-link>
    </div><pds-link href="#" id="testTwo" class="pds-c-card__element-with-hover-state">This is a PDS link</pds-link><div slot="footer">
    <pds-link href="#" id="testThree" class="pds-c-card__element-with-hover-state">This is a PDS tag link</pds-link>
  </div></pds-card>`,
    )) as PdsCard;

    expect(configuredElement.hasNoSlottedLinkContent).toBe(false);

    expect(
      configuredElement
        .querySelector('#test')
        ?.classList.contains('pds-c-card__element-with-hover-state'),
    ).toBeTruthy();

    expect(
      configuredElement
        .querySelector('#testTwo')
        ?.classList.contains('pds-c-card__element-with-hover-state'),
    ).toBeTruthy();

    expect(
      configuredElement
        .querySelector('#testThree')
        ?.classList.contains('pds-c-card__element-with-hover-state'),
    ).toBeTruthy();
  });

  it('should render nothing when href is provided, there is no slotted link, and no aria label', async () => {
    const consoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    const configuredElement = await fixture('<pds-card href="#"/>');

    const card = configuredElement.shadowRoot?.querySelector('.pds-c-card');

    expect(card).toBeFalsy();
    expect(console.error).toHaveBeenCalledTimes(1);

    consoleError.mockRestore();
  });

  it('should console warn when an ariaLabel is provided but there is no clickable', async () => {
    const consoleError = jest
      .spyOn(console, 'warn')
      .mockImplementation(() => {});

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const configuredElement = await fixture(
      '<pds-card ariaLabel="Test"></pds-card>',
    );

    expect(consoleError).toBeCalledTimes(1);

    consoleError.mockRestore();
  });

  it('should console warn when a target is provided but there is no href', async () => {
    const consoleError = jest
      .spyOn(console, 'warn')
      .mockImplementation(() => {});

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const configuredElement = await fixture(
      '<pds-card target="_blank"></pds-card>',
    );

    expect(consoleError).toBeCalledTimes(1);

    consoleError.mockRestore();
  });

  it('should console warn when a direction is horizontal and centerContentVertically is provided', async () => {
    const consoleError = jest
      .spyOn(console, 'warn')
      .mockImplementation(() => {});

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const configuredElement = await fixture(
      '<pds-card direction="horizontal" centerContentVertically></pds-card>',
    );

    expect(consoleError).toBeCalledTimes(1);

    consoleError.mockRestore();
  });

  it('should return an empty string from getSummaryText if textContent is null', async () => {
    const typedElement = element as PdsCard;

    expect(typedElement.getSummaryText(null)).toEqual('');
  });

  it('adds the appropriate class when centerContentVertically prop is passed in', async () => {
    const element1 = await fixture(
      `<pds-card
      variant="default"
      centerContentVertically="true">
      <div slot="header">
        <placeholder-element style="--placeholder-element-margin-bottom: 0;"
          >Card header slot</placeholder-element
        >
      </div>
      <placeholder-element style="--placeholder-element-margin-bottom: 0;"
        >Card body slot</placeholder-element
      >
      <div slot="footer">
        <placeholder-element style="--placeholder-element-margin-bottom: 0;"
          >Card footer slot</placeholder-element
        >
      </div>
    </pds-card>`,
    );
    const card1 = element1.shadowRoot?.querySelector(
      '.pds-c-card',
    ) as HTMLElement;

    expect(card1?.classList.contains('pds-c-card--center-vertically')).toBe(
      true,
    );
  });

  it('does not add the appropriate class when centerContentVertically prop is passed in when direction is horizontal', async () => {
    const configuredElement = await fixture(
      '<pds-card direction="horizontal" centerContentVertically></pds-card>',
    );
    const card = configuredElement.shadowRoot?.querySelector(
      '.pds-c-card',
    ) as HTMLElement;

    expect(card?.classList.contains('pds-c-card--center-vertically')).toBe(
      false,
    );
  });
});
