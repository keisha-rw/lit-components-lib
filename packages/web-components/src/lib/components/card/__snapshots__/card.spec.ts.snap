// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card unit tests populates the footer slot when provided 1`] = `
<pds-card>
  <div
    slot="footer"
  >
    Footer text and stuff
  </div>
</pds-card>
`;

exports[`Card unit tests populates the header slot when provided 1`] = `
<pds-card>
  <div
    slot="header"
  >
    header text and stuff
  </div>
</pds-card>
`;
